pnpm run milady start crashes with Segmentation Fault on Bun 1.3.9 (Linux x64)

### Description

Running pnpm run milady start causes the process to crash with a Segmentation Fault after ~1 minute of running. The process uses Bun as its runtime and fails consistently.

### Steps to Reproduce

1. Clone the repo and install dependencies: pnpm install --no-strict-peer-dependencies
2. Run pnpm build
3. Run pnpm run milady start
4. Wait approximately 1 minute

### Expected Behavior

The server runs stably and keeps serving requests indefinitely.

### Actual Behavior

The process crashes with a Segmentation Fault:

panic: Segmentation fault at address 0x796CC1907440
oh no: Bun has crashed. This indicates a bug in Bun, not your code.
Bun v1.3.9 (cf6cdbbb) Linux x64
Linux Kernel v6.14.0 | glibc v2.39
RSS: 0.99GB | Peak: 0.71GB | Commit: 0.99GB

### Environment

Linux

### Node Version

22.22.0

### Model Provider

_No response_

### Logs

```shell
Bun crash report: https://bun.report/1.3.9/la1cf6cdbbinGugooC4qktxE+ypR+stGA2428Bh8l+t+B
```

### Screenshots

_No response_

--- a/scripts/run-node-runtime.mjs
+++ b/scripts/run-node-runtime.mjs
@@ -1,0 +1,79 @@
+const KNOWN_UNSTABLE_BUN_LINUX = /^1\.3\.9(?:$|[-+].*)/;
+
+/**
+ * Bun 1.3.9 has known Linux segfault reports in long-running workloads.
+ * Prefer Node by default for this one runtime/version combination.
+ */
+export function isKnownUnstableBunOnLinux({ platform, bunVersion }) {
+  return (
+    platform === "linux" &&
+    typeof bunVersion === "string" &&
+    KNOWN_UNSTABLE_BUN_LINUX.test(bunVersion)
+  );
+}
+
+/**
+ * Runtime selection priority:
+ * 1) Explicit MILADY_RUNTIME override (bun|node)
+ * 2) Safety fallback for known unstable Bun/Linux combo
+ * 3) Default to bun
+ */
+export function chooseMiladyRuntime({
+  requestedRuntime,
+  platform,
+  bunVersion,
+}) {
+  const normalized = requestedRuntime?.trim().toLowerCase();
+  if (normalized === "bun" || normalized === "node") {
+    return { runtime: normalized, warning: null };
+  }
+
+  if (isKnownUnstableBunOnLinux({ platform, bunVersion })) {
+    return {
+      runtime: "node",
+      warning:
+        "Detected Bun 1.3.9 on Linux (known segfault risk). Defaulting runtime to Node.js.",
+    };
+  }
+
+  return { runtime: "bun", warning: null };
+}
+
+export function resolveNodeExecPath({
+  currentExecPath,
+  platform,
+  explicitNodePath,
+}) {
+  const explicit = explicitNodePath?.trim();
+  if (explicit) {
+    return explicit;
+  }
+
+  const normalized =
+    platform === "win32"
+      ? (currentExecPath ?? "").toLowerCase()
+      : (currentExecPath ?? "");
+  const looksLikeBun = /(?:^|[\\/])bun(?:\.exe)?$/.test(normalized);
+
+  if (!looksLikeBun && normalized.length > 0) {
+    return currentExecPath;
+  }
+
+  return platform === "win32" ? "node.exe" : "node";
+}
+
+export function resolveRuntimeExecPath({
+  runtime,
+  currentExecPath,
+  platform,
+  explicitNodePath,
+}) {
+  if (runtime === "bun") {
+    return "bun";
+  }
+  return resolveNodeExecPath({
+    currentExecPath,
+    platform,
+    explicitNodePath,
+  });
+}

--- a/scripts/run-node-runtime.test.ts
+++ b/scripts/run-node-runtime.test.ts
@@ -1,0 +1,113 @@
+import { describe, expect, it } from "vitest";
+import {
+  chooseMiladyRuntime,
+  isKnownUnstableBunOnLinux,
+  resolveNodeExecPath,
+  resolveRuntimeExecPath,
+} from "./run-node-runtime.mjs";
+
+describe("run-node runtime selection", () => {
+  it("detects Bun 1.3.9 on Linux as unstable", () => {
+    expect(
+      isKnownUnstableBunOnLinux({ platform: "linux", bunVersion: "1.3.9" }),
+    ).toBe(true);
+    expect(
+      isKnownUnstableBunOnLinux({
+        platform: "linux",
+        bunVersion: "1.3.9-hotfix",
+      }),
+    ).toBe(true);
+    expect(
+      isKnownUnstableBunOnLinux({ platform: "darwin", bunVersion: "1.3.9" }),
+    ).toBe(false);
+    expect(
+      isKnownUnstableBunOnLinux({ platform: "linux", bunVersion: "1.3.10" }),
+    ).toBe(false);
+  });
+
+  it("honors explicit runtime overrides", () => {
+    expect(
+      chooseMiladyRuntime({
+        requestedRuntime: "node",
+        platform: "linux",
+        bunVersion: "1.3.9",
+      }),
+    ).toEqual({ runtime: "node", warning: null });
+
+    expect(
+      chooseMiladyRuntime({
+        requestedRuntime: "bun",
+        platform: "linux",
+        bunVersion: "1.3.9",
+      }),
+    ).toEqual({ runtime: "bun", warning: null });
+  });
+
+  it("falls back to node for linux Bun 1.3.9 when runtime is not set", () => {
+    const selected = chooseMiladyRuntime({
+      requestedRuntime: undefined,
+      platform: "linux",
+      bunVersion: "1.3.9",
+    });
+    expect(selected.runtime).toBe("node");
+    expect(selected.warning).toContain("Bun 1.3.9");
+  });
+
+  it("defaults to bun otherwise", () => {
+    expect(
+      chooseMiladyRuntime({
+        requestedRuntime: undefined,
+        platform: "linux",
+        bunVersion: "1.3.10",
+      }),
+    ).toEqual({ runtime: "bun", warning: null });
+  });
+});
+
+describe("run-node executable resolution", () => {
+  it("uses node on PATH when current exec is bun", () => {
+    expect(
+      resolveNodeExecPath({
+        currentExecPath: "/Users/home/.bun/bin/bun",
+        platform: "linux",
+      }),
+    ).toBe("node");
+  });
+
+  it("keeps current exec path when it already points to node", () => {
+    expect(
+      resolveNodeExecPath({
+        currentExecPath: "/usr/bin/node",
+        platform: "linux",
+      }),
+    ).toBe("/usr/bin/node");
+  });
+
+  it("prefers explicit node path override", () => {
+    expect(
+      resolveNodeExecPath({
+        currentExecPath: "/Users/home/.bun/bin/bun",
+        platform: "linux",
+        explicitNodePath: "/custom/node",
+      }),
+    ).toBe("/custom/node");
+  });
+
+  it("resolves runtime executable for bun and node", () => {
+    expect(
+      resolveRuntimeExecPath({
+        runtime: "bun",
+        currentExecPath: "/usr/bin/node",
+        platform: "linux",
+      }),
+    ).toBe("bun");
+
+    expect(
+      resolveRuntimeExecPath({
+        runtime: "node",
+        currentExecPath: "/Users/home/.bun/bin/bun",
+        platform: "linux",
+      }),
+    ).toBe("node");
+  });
+});

--- a/scripts/run-node.mjs
+++ b/scripts/run-node.mjs
@@ -3,6 +3,10 @@ import { spawn } from "node:child_process";
 import fs from "node:fs";
 import path from "node:path";
 import process from "node:process";
+import {
+  chooseMiladyRuntime,
+  resolveRuntimeExecPath,
+} from "./run-node-runtime.mjs";
 
 const args = process.argv.slice(2);
 const env = { ...process.env };
@@ -113,9 +117,20 @@ const logRunner = (message) => {
 const RESTART_EXIT_CODE = 75;
 
 const runNode = () => {
-  // Eliza MIGRATION: Use bun for faster startup and better TypeScript support
-  const runtime = process.env.MILADY_RUNTIME || "bun";
-  const execPath = runtime === "bun" ? "bun" : process.execPath;
+  const { runtime, warning } = chooseMiladyRuntime({
+    requestedRuntime: process.env.MILADY_RUNTIME,
+    platform: process.platform,
+    bunVersion: process.versions?.bun,
+  });
+  if (warning) {
+    logRunner(`${warning} Set MILADY_RUNTIME=bun to force Bun runtime.`);
+  }
+  const execPath = resolveRuntimeExecPath({
+    runtime,
+    currentExecPath: process.execPath,
+    platform: process.platform,
+    explicitNodePath: process.env.MILADY_NODE_PATH,
+  });
   const nodeProcess = spawn(execPath, ["milady.mjs", ...args], {
     cwd,
     env,

